core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -106,6 +106,14 @@ protected abstract TableScan newRefinedScan(
       TableOperations ops, Snapshot snapshot, Expression rowFilter,
       boolean ignoreResiduals, boolean caseSensitive, boolean colStats);
 
+  /**
+   * The {@link #filter(Expression)} is based on the table records. So the data table scan need select all of the
+   * filter columns, other scans don't need to.
+   */
+  protected boolean isDataTableScan() {
```

Comment:
I'm not sure I understand why this is added. Shouldn't the filter (which is unbound) be bound to the scan's schema and not the underlying table's schema?